### PR TITLE
[1.11] Multiple Persistence Units

### PR DESCRIPTION
**Pull request by mjurc**
Adding a test application verifying that configuration of multiple persistence units within application.properties works.

RFE: https://issues.redhat.com/browse/QUARKUS-715
Upstream PR: https://github.com/quarkus-qe/quarkus-openshift-test-suite/pull/242

### CODE REVIEW

**Comment by Sgitario** [comment]
run tests

**Comment by Sgitario** [comment]
run tests

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/ApplicationExceptionMapper.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+
+@Provider
+public class ApplicationExceptionMapper implements ExceptionMapper<Exception> {
+    @Override
+    public Response toResponse(Exception exception) {
+        int code = 500;
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/ApplicationExceptionMapper.java:15]: Avoid magic numbers ( numeric literal value). Use a constant instead.

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/VegetableResource.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    @GET
+    public List<Vegetable> getAll() {
+        return Vegetable.listAll(Sort.by("name"));
+    }
+
+    @GET
+    @Path("/{id}")
+    public Vegetable get(@PathParam("id") Long id) {
+        Vegetable vegetable = Vegetable.findById(id);
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/VegetableResource.java:43]: Consider using a final local variable. Or replace this statement with an Optional
Example:
`Optional.ofNullable(Vegetable.findById(id)).orElseThrow(()->new NotFoundException("vegetable '" + id + "' not found"));`

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/VegetableResource.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+        vegetable.persist();
+        return Response.ok(vegetable).status(201).build();
+    }
+
+    @PUT
+    @Path("/{id}")
+    @Transactional
+    public Vegetable update(@PathParam("id") Long id, @Valid Vegetable newVegetable) {
+        Vegetable vegetable = Vegetable.findById(id);
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/VegetableResource.java:65]: Consider using a final local variable. Or replace this statement with an Optional
Example:
Optional.ofNullable(Vegetable.findById(id)).orElseThrow(()->new NotFoundException("vegetable '" + id + "' not found"));

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/VegetableResource.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+        vegetable.name = newVegetable.name;
+        return vegetable;
+    }
+
+    @DELETE
+    @Path("/{id}")
+    @Transactional
+    public Response delete(@PathParam("id") Long id) {
+        Vegetable vegetable = Vegetable.findById(id);
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/VegetableResource.java:78]: Consider using a final local variable. Or replace this statement with an Optional
Example:
Optional.ofNullable(Vegetable.findById(id)).orElseThrow(()->new NotFoundException("vegetable '" + id + "' not found"));

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+        fruit.persist();
+        return Response.ok(fruit).status(201).build();
+    }
+
+    @PUT
+    @Path("/{id}")
+    @Transactional
+    public Fruit update(@PathParam("id") Long id, @Valid Fruit newFruit) {
+        Fruit fruit = Fruit.findById(id);
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java:65]: Consider using a final local variable. Or replace this statement with an Optional
Example:
Optional.ofNullable(Vegetable.findById(id)).orElseThrow(()->new NotFoundException("vegetable '" + id + "' not found"));

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    @GET
+    public List<Fruit> getAll() {
+        return Fruit.listAll(Sort.by("name"));
+    }
+    
+    @GET
+    @Path("/{id}")
+    public Fruit get(@PathParam("id") Long id) {
+        Fruit fruit = Fruit.findById(id);
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java:43]: Consider using a final local variable. Or replace this statement with an Optional
Example:
Optional.ofNullable(Vegetable.findById(id)).orElseThrow(()->new NotFoundException("vegetable '" + id + "' not found"));

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+        fruit.name = newFruit.name;
+        return fruit;
+    }
+
+    @DELETE
+    @Path("/{id}")
+    @Transactional
+    public Response delete(@PathParam("id") Long id) {
+        Fruit fruit = Fruit.findById(id);
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java:78]: Consider using a final local variable. Or replace this statement with an Optional
Example:
Optional.ofNullable(Vegetable.findById(id)).orElseThrow(()->new NotFoundException("vegetable '" + id + "' not found"));

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/ValidationExceptionMapper.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        ObjectMapper mapper = new ObjectMapper();
+        ArrayNode errors = mapper.createArrayNode();
+
+        for (ConstraintViolation<?> constraintViolation : exception.getConstraintViolations()) {
+            errors.addObject()
+                    .put("path", constraintViolation.getPropertyPath().toString())
+                    .put("message", constraintViolation.getMessage());
+        }
+
+        return Response.status(422)
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/ValidationExceptionMapper.java:26]: Avoid magic numbers used a constant instead.

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/VegetableResource.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            throw new NotFoundException("vegetable '" + id + "' not found");
+        }
+        return vegetable;
+    }
+
+    @POST
+    @Transactional
+    public Response create(@Valid Vegetable vegetable) {
+        if (vegetable.id != null) {
+            throw new ClientErrorException("unexpected ID in request", 422);
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/VegetableResource.java:54]: Avoid magic numbers used a constant instead.

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            throw new NotFoundException("fruit '" + id + "' not found");
+        }
+        return fruit;
+    }
+
+    @POST
+    @Transactional
+    public Response create(@Valid Fruit fruit) {
+        if (fruit.id != null) {
+            throw new ClientErrorException("unexpected ID in request", 422);
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java:54]: Avoid magic numbers used a constant instead.

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        fruit.persist();
+        return Response.ok(fruit).status(201).build();
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java:58]: Avoid magic numbers used a constant instead.

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/VegetableResource.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        vegetable.persist();
+        return Response.ok(vegetable).status(201).build();
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/VegetableResource.java:58]: Avoid magic numbers used a constant instead.

File: sql-db/multiple-pus/src/test/java/io/quarkus/ts/openshift/sqldb/multiplepus/AbstractMultiplePersistenceUnitTest.java [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+                .post("/vegetable")
+                .then()
+                .statusCode(201)
+                .body("id", equalTo(8))
+                .body("name", equalTo("Eggplant"));
+
+        when()
+                .get("/vegetable/8")
+                .then()
+                .statusCode(200)
```
Comment (pjgg) [inline sql-db/multiple-pus/src/test/java/io/quarkus/ts/openshift/sqldb/multiplepus/AbstractMultiplePersistenceUnitTest.java:231]: Consider using a constant to all HTTP Status codes.

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/ValidationExceptionMapper.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+import javax.ws.rs.ext.Provider;
+
+@Provider
+public class ValidationExceptionMapper implements ExceptionMapper<ConstraintViolationException> {
+    @Override
+    public Response toResponse(ConstraintViolationException exception) {
+        ObjectMapper mapper = new ObjectMapper();
+        ArrayNode errors = mapper.createArrayNode();
+
+        for (ConstraintViolation<?> constraintViolation : exception.getConstraintViolations()) {
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/ValidationExceptionMapper.java:20]: Consider to declare this variable `constraintViolation` final

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/ValidationExceptionMapper.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+
+@Provider
+public class ValidationExceptionMapper implements ExceptionMapper<ConstraintViolationException> {
+    @Override
+    public Response toResponse(ConstraintViolationException exception) {
+        ObjectMapper mapper = new ObjectMapper();
+        ArrayNode errors = mapper.createArrayNode();
```
Comment (pjgg) [inline sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/ValidationExceptionMapper.java:18]: these two variables could be final

**Review by pjgg — CHANGES_REQUESTED** [review]

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/ValidationExceptionMapper.java (hunk 8 above)
Comment (pjgg) [reply sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/ValidationExceptionMapper.java:26]: Try to use something like this for Http Status code: `Response.Status.NOT_FOUND`

File: sql-db/multiple-pus/src/main/resources/import-fruits.sql [hunk 16]
```diff
@@ -0,0 +1,8 @@
+INSERT INTO fruit (id, name) VALUES (1, 'Apple');
+INSERT INTO fruit (id, name) VALUES (2, 'Pear');
+INSERT INTO fruit (id, name) VALUES (3, 'Strawberry');
+INSERT INTO fruit (id, name) VALUES (4, 'Plum');
+INSERT INTO fruit (id, name) VALUES (5, 'Cherry');
+INSERT INTO fruit (id, name) VALUES (6, 'Berry');
+INSERT INTO fruit (id, name) VALUES (7, 'Cranberry');
+UPDATE hibernate_sequence SET next_val = 8;
```
Comment (jsmrcka) [inline sql-db/multiple-pus/src/main/resources/import-fruits.sql:8]: Can we use `nextval` in the insert statements?

File: sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java (hunk 6 above)
Comment (Ladicek) [reply sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java:43]: The existing code is just fine.

Comment (pjgg) [reply sql-db/multiple-pus/src/main/java/io/quarkus/ts/openshift/sqldb/multiplepus/FruitResource.java:43]: this is a suggestion!

**Comment by rsvoboda** [comment]
@mjurc this need to move ahead asap!

File: sql-db/multiple-pus/src/main/resources/import-fruits.sql (hunk 16 above)
Comment (mjurc) [reply sql-db/multiple-pus/src/main/resources/import-fruits.sql:8]: No, `mariadb` doesn't like it much

**Review by pjgg — APPROVED** [review]

**Comment by mjurc** [comment]
@rsvoboda @pjgg  The unit tests fail when run against Quarkus master right now. I've been doing a bit of investigation and workarounding - it seems there's something wrong with initialising the `hibernate_seq` sequence for `import*.sql` scripts for H2. This seems like a regression, but I have not found a simplistic way to work around this.

We can either disable the tests until I find workaround/the fix is implemented (this would be kinda shame though - the tests work for 1.11 and 1.12), or proceed with this MR for now and add this as an issue for investigation.

**Comment by mjurc** [comment]
@rsvoboda @pjgg I skipped the test for main branch with the reported bug. This should be ready to merge now.

**Comment by pjgg** [comment]
LGTM.

**Pull request merged**